### ISSUE TITLE
Add --cov-omit option

### ISSUE THREAD

**Issue by roipoussiere** [pytest-dev/pytest-cov#373]
Similar to `--omit` option in [coverage.py](https://coverage.readthedocs.io/en/latest/source.html#execution), this option could be used to ignore some specific files. 

Ignoring files can be done with a `.coveragerc` file, but many developers don't want to add an other file in their project root folder just to ignore one or two folder during the coverage analysis.

related to #193

**Comment by nedbat** [comment]
Coverage will read settings from setup.cfg, or tox.ini, or pyproject.toml files. There's no need to add another option to an already cluttered command line.

**Comment by roipoussiere** [comment]
Yes but it can be annoying to create a configuration file only to add a simple option to the command.

**Comment by nedbat** [comment]
@roipoussiere You don't have any of these files? setup.cfg, tox.ini, pyproject.toml?

**Comment by roipoussiere** [comment]
No, any of them. I would like to keep my root tree as clean as possible.

**Comment by nedbat** [comment]
I understand that impulse. We would like to keep our plugin as simple as possible.

**Comment by MichaelKim0407** [comment]
I think this suggestion is completely valid. Sometimes I just want to try a command to see if something works, without creating a config file. That's the whole point of command line arguments.

If I need to use the same configuration repeatedly, I'll know myself to create a config file. I don't need you to dictate that I should always create a config file.

And it's not like this suggestion is asking for a very complex argument. Whatever you already support in the config file, just put it in the command line. With your logic (to keep it simple) you may as well argue that your plugin shouldn't support any command line argument at all.

**Comment by The-Compiler** [comment]
@nedbat How would you feel about having a general "coverage option" commandline argument, similar to pytest's `-o` / `--override-ini`? That way, people could still do something like `--cov-opt run.omit resources.py` etc., but the plugin doesn't need to play catch up if there are new arguments/options in coverage.py?

**Comment by MichaelKim0407** [comment]
@The-Compiler I think your suggestion somewhat misses the point:

1. The reason for this feature request is that _I don't want to create a config file_

2. It's not playing catch up. `pytest-cov` already supports it. It's just a matter of adding it in the command line options.

**Comment by The-Compiler** [comment]
> @The-Compiler I think your suggestion somewhat misses the point

It does?

>  The reason for this feature request is that _I don't want to create a config file_

Which is why I proposed an option which lets you set coverage config options without having a config file.

> It's not playing catch up. `pytest-cov` already supports it. It's just a matter of adding it in the command line options.

How so? From what I can tell, pytest-cov does not handle `omit` in any way. Having to play catch-up because the plugin is rather complex has been a problem in the past (see #337 for a longer discussion), hence my proposal of adding an option which works in a more general way and doesn't need the plugin to update when there are changes in coverage.py.

**Comment by MichaelKim0407** [comment]
@The-Compiler ok nevermind, my bad

**Comment by roipoussiere** [comment]
The discussion continues here with good arguments but the issue is still closed. Is it possible to re-open it?

**Comment by ionelmc** [comment]
It is certainly possible but whomever reopens it should pledge to maintain an increasing cornucopia of options and switches. That person isn't me and I don't see anyone offering their time when clearly there are alternatives to another cli option.

**Comment by MichaelKim0407** [comment]
In that case I think the issue should be reopened so that anyone interested
can pick it up. Closed issue means that the maintainer of this project
isn't interested.

In other words, if the maintainer of this project is open to this feature
being implemented, the issue should be open.

**Comment by ionelmc** [comment]
I'm not interested. I was only alluding that maintenance is scarce.

**Comment by The-Compiler** [comment]
@ionelmc How do you feel about my proposal in https://github.com/pytest-dev/pytest-cov/issues/373#issuecomment-603779658? That could possibly even replace any existing commandline options which just get passed through to coverage.py currently. However, it means coverage.py would need some kind of API to pass additional options (while still reading its config file).

**Comment by ionelmc** [comment]
I guess it could be acceptable if that makes it possible to remove other cli arguments.

**Comment by philipjames44** [comment]
This would be a really nice QOL feature

**Comment by cygnus2048** [comment]
Yeah, I've been waiting for this for several years now and still nothing.  Instead in every one of my projects I have to hack the shared config file that all my projects use with multiple sed commands.  Sad.

**Comment by mikeloomisgg** [comment]
I believe the dev spoke past the original ask here. If you came to this thread looking for a solution to be able to specify both source files to include in coverage, and also paths to omit from coverage, you should use the [tool.coverage] metadata in your pyproject.toml. **You do not need to create a .coveragerc file**, and doing such would obviously not be canonical in modern python.

Here is an example for adding omit options to your pyproject.toml:
```toml
[tool.pytest.ini_options]
testpaths = ["tests"]
norecursedirs = ["dist", "build", "dev", "typings"]
addopts = ["--cov=./"]

[tool.coverage.run]
source = ["./*"]
omit = ["dev/*", "tests/*"]
```

This satisfies canonical python projects, since all of them should be using a single pyproject.toml to define their project's metadata.

**Comment by cygnus2048** [comment]
I have a common coverage config file that is copied into each project during the build.  As one of the final steps of the build I run multiple sed commands to add to the "omit" section of that config file.  It works, but less than desirable.  Having a command line arg would make this much simpler and cleaner.  Thanks.

**Comment by mattvonrocketstein** [comment]
for projects that have multiple test-suites (say units vs smoke-testing) using a base coverage config file + a few extra options would be very nice to have.  

more specifically, i want to do things like run unit-tests and omit coverage for `__main__.py`, because those are covered in a smoke-test suite elsewhere.  (suite-coverage will be combined later if all suites are running, but if my work is more focused, it's good to avoid the screen-clutter and cognitive overhead of parsing the irrelevant coverage.  plus it's bring down the percentage :)

afaik, currently the best solution here is to have multiple .coveragerc's where the vast majority of the content is redundant and then has to be kept in sync.  i imagine this scenario is fairly common; just wanted to share that everyone who is interested in this is not trying to completely avoid config files and put everything on a CLI invocation.  i have and enjoy config files!, but i don't want half a dozen for simple tasks, so base configs + CLI overrides seems like the way to go.

**Comment by aarz-snl** [comment]
This needs to be added. Kinda upsetting it was shutdown so quickly. Especially needed in ci/cd environments when you have hundreds of repos sourcing a single yml. Wouldn't it make sense to have an option to make ONE change to the yaml that instantly is applied to all repos?

**Comment by nedbat** [comment]
Just to be clear, I am not a maintainer of pytest-cov. I maintain coverage.py, the engine underpinning this plugin.

There are many ways to configure coverage.py.  For example, you can use environment variables in its configuration file, so "maybe an env-var" is already there.

> afaik, currently the best solution here is to have multiple .coveragerc's where the vast majority of the content is redundant and then has to be kept in sync.

Look into using environment variables in the .coveragerc files. @mattvonrocketstein If that doesn't work for you, can you tell us more about your specifics?

I'd like to understand more about some of the unusual situations people are describing:

> ci/cd environments when you have hundreds of repos sourcing a single yml.

@aarz-snl can you tell us more about this?  If there's a single yml file, where will the individualized command-line arguments for pytest-cov be specified?  I guess I don't know which yml file you mean.

> I have a common coverage config file that is copied into each project during the build. As one of the final steps of the build I run multiple sed commands to add to the "omit" section of that config file. It works, but less than desirable. Having a command line arg would make this much simpler and cleaner.

@cygnus2048 can you explain more? Why is it easy to have different command line arguments for each project while it's hard to have different configuration files?  Also, have you looked into using environment variables in the configuration file?

I don't mean to be difficult about this.  I want to make sure people know about the power and flexibility that is already available to them.  I also want to understand where there isn't enough power.

**Comment by andrewwarz** [comment]
@nedbat For my situation with CI/CD:

In GitLab I use template CI/CD configurations. So, all of my repos underneath it 'include' them. 

```
Template Repo
     │
     │ includes
     ▼
┌─────────┐
│ Repo A  │
├─────────┤
│ Repo B  │
├─────────┤
│ Repo C  │
├─────────┤
│ Repo D  │
└─────────┘

One template → Many repos
```

 So, the suggestion kept coming up to just add a .coveragerc file -- I prefer to AVOID having to add more and more configuration files to my repos that I have to maintain as much as possible. I got around the issue by just dynamically generating a .coveragerc file each time the job runs.

**Comment by nedbat** [comment]
@andrewwarz thanks, details like this are really helpful.  How would you use a `--cov-omit` option if it were available? In your templated CI/CD configurations, where is it easy to provide that option differently for each repo?

Another question: what files are people trying to omit from coverage, and why? I'm not saying you shouldn't, just trying to understand the whole problem.

**Comment by andrewwarz** [comment]
@nedbat testing files used during pytests which dont need coverage reports, dynamically generated files like _version.py created by setuptools-scm

**Comment by nedbat** [comment]
@andrewwarz:
> How would you use a --cov-omit option if it were available? In your templated CI/CD configurations, where is it easy to provide that option differently for each repo?

**Comment by andrewwarz** [comment]
@nedbat I ended up using the script section of the template job to do something like this:

```yaml
  script:
    - |
        echo "[run]" > .coveragerc
        echo "omit = _version.py" >> .coveragerc
        echo some other stuff redacted
```

When I first started I spent some time hoping it could be done like this

```
`python -m pytest --cov=PROJECT_PATH --cov-omit=__version.py, etc, etc`
```

Ultimately, the workaround was fine -- just clunky, and more scripting than I wanted to do to complete the task
